### PR TITLE
[Merged by Bors] - feat: port Order.OmegaCompletePartialOrder

### PR DESCRIPTION
**Pull request by rwbarton**
- [x] depends on: #1397

### CODE REVIEW

**Comment by rwbarton** [comment]
https://github.com/leanprover/lean4/issues/1937 makes this too unpleasant for now.

**Comment by hrmacbeth** [comment]
> [leanprover/lean4#1937](https://github.com/leanprover/lean4/issues/1937) makes this too unpleasant for now.

Hopefully fixed after the bump to nightly-2023-01-06.

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#1397
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#1397~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Order/OmegaCompletePartialOrder.lean [hunk 1]
```diff
@@ -0,0 +1,934 @@
...
+
+namespace CompleteLattice
+
+variable {α β : Type _} [OmegaCompletePartialOrder α] [CompleteLinearOrder β]
+
+theorem inf_continuous (f g : α →o β) (hf : Continuous f) (hg : Continuous g) :
+    Continuous (f ⊓ g) := by
+  refine' fun c => eq_of_forall_ge_iff fun z => _
+  simp only [inf_le_iff, hf c, hg c, ωSup_le_iff, ← forall_or_left, ← forall_or_right,
+    Function.comp_apply, Chain.map_coe, OrderHom.orderBot] -- something broken here
```
Comment (Ruben-VandeVelde) [inline Mathlib/Order/OmegaCompletePartialOrder.lean:533]: It seems like `inf_le_iff` no longer applies. Trying to force it gets me to a `@SemilatticeInf.toHasInf β Lattice.toSemilatticeInf  = @Lattice.toHasInf β LinearOrder.toLattice` goal.

Also the lemma now known as `OrderHom.instHasInfOrderHomToPreorderToPartialOrder_inf_coe` is missing from the simp call

Comment (ChrisHughes24) [reply Mathlib/Order/OmegaCompletePartialOrder.lean:533]: I fixed this problem with `inf_le_iff` no longer applying.

File: Mathlib/Order/OmegaCompletePartialOrder.lean [hunk 2]
```diff
@@ -0,0 +1,942 @@
...
+variable {f}
+
+theorem mem_map (x : α) : x ∈ c → f x ∈ Chain.map c f :=
+  fun ⟨i, h⟩ => ⟨i, h.symm ▸ rfl⟩
+#align omega_complete_partial_order.chain.mem_map OmegaCompletePartialOrder.Chain.mem_map
+
+theorem exists_of_mem_map {b : β} : b ∈ c.map f → ∃ a, a ∈ c ∧ f a = b :=
+  fun ⟨i, h⟩ => ⟨c i, ⟨i, rfl⟩, h.symm⟩
+#align
+  omega_complete_partial_order.chain.exists_of_mem_map OmegaCompletePartialOrder.Chain.exists_of_mem_map
```
Comment (jcommelin) [inline Mathlib/Order/OmegaCompletePartialOrder.lean:128]:
```suggestion
  omega_complete_partial_order.chain.exists_of_mem_map
  OmegaCompletePartialOrder.Chain.exists_of_mem_map
```

File: Mathlib/Order/OmegaCompletePartialOrder.lean [hunk 3]
```diff
@@ -537,12 +533,11 @@ variable {α β : Type _} [OmegaCompletePartialOrder α] [CompleteLinearOrder β
...
   refine' fun c => eq_of_forall_ge_iff fun z => _
-  simp only [inf_le_iff, hf c, hg c, ωSup_le_iff, ← forall_or_left, ← forall_or_right,
-    Function.comp_apply, chain.map_coe, OrderHom.has_inf_inf_coe]
-  exact
-    ⟨fun h _ => h _ _, fun h i j =>
-      (h (max i j)).imp (le_trans <| f.mono <| c.mono <| le_max_left _ _)
-        (le_trans <| g.mono <| c.mono <| le_max_right _ _)⟩
+  --Porting note: bad non-terminal `simp`, but replacing with the output of `simp?` doesn't work
+  simp [inf_le_iff, hf c, hg c, ωSup_le_iff, ←forall_or_left, ←forall_or_right,
+    Chain.map_coe, OrderHom.coe_inf, Function.comp]
```
Comment (ericrbg) [inline Mathlib/Order/OmegaCompletePartialOrder.lean:538]:
```suggestion
  simp only [inf_le_iff, hf c, hg c, ωSup_le_iff, ←forall_or_left, ←forall_or_right,
             Chain.map_coe, OrderHom.coe_inf, ge_iff_le, Pi.inf_apply, Function.comp]
```

fixes the non-terminal simp

**Review by ericrbg — APPROVED** [review]
otherwise lgtm

File: Mathlib/Order/OmegaCompletePartialOrder.lean [hunk 4]
```diff
@@ -0,0 +1,942 @@
...
+attribute [-simp] Part.bind_eq_bind Part.map_eq_map
+
+open Classical
+
+namespace OrderHom
+
+variable {α : Type _} {β : Type _} {γ : Type _}
+variable [Preorder α] [Preorder β] [Preorder γ]
+
+/-- `part.bind` as a monotone function -/
```
Comment (jcommelin) [inline Mathlib/Order/OmegaCompletePartialOrder.lean:72]:
```suggestion
/-- `Part.bind` as a monotone function -/
```

File: Mathlib/Order/OmegaCompletePartialOrder.lean [hunk 5]
```diff
@@ -0,0 +1,942 @@
...
+instance : Membership α (Chain α) :=
+  ⟨fun a (c : ℕ →o α) => ∃ i, a = c i⟩
+
+variable (c c' : Chain α)
+variable (f : α →o β)
+variable (g : β →o γ)
+
+instance : LE (Chain α) where le x y := ∀ i, ∃ j, x i ≤ y j
+
+/-- `map` function for `chain` -/
```
Comment (jcommelin) [inline Mathlib/Order/OmegaCompletePartialOrder.lean:113]:
```suggestion
/-- `map` function for `Chain` -/
```

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**